File: cloud/tests/unit/test_middleware.py
"""Unit tests for client certificate middleware."""

import json

from starlette.testclient import TestClient

from uplink.middleware import extract_client_subject
from uplink.app import app


client = TestClient(app)


class TestExtractClientSubjectViaLWA:
    """Cert extraction from Lambda Web Adapter x-amzn-request-context header."""

    def test_extracts_subject_from_request_context(self) -> None:
        ctx = {
            "identity": {
                "clientCert": {
                    "subjectDN": "CN=device-001,O=Supervictor",
                }
            }
        }
        response = client.get(
            "/", headers={"x-amzn-request-context": json.dumps(ctx)}
        )
        assert response.status_code == 200
        assert response.json()["client_subject"] == "CN=device-001,O=Supervictor"

    def test_returns_none_for_malformed_context(self) -> None:
        response = client.get(
            "/", headers={"x-amzn-request-context": "not-json"}
        )
        assert response.status_code == 200
        assert "client_subject" not in response.json()

    def test_returns_none_for_missing_cert_in_context(self) -> None:
        ctx = {"identity": {}}
        response = client.get(
            "/", headers={"x-amzn-request-context": json.dumps(ctx)}
        )
        assert response.status_code == 200
        assert "client_subject" not in response.json()


class TestExtractClientSubjectViaProxy:
    """Cert extraction from reverse proxy header."""

    def test_extracts_subject_from_ssl_header(self) -> None:
        response = client.get(
            "/",
            headers={"x-ssl-client-subject-dn": "CN=device-002,O=Supervictor"},
        )
        assert response.status_code == 200
        assert response.json()["client_subject"] == "CN=device-002,O=Supervictor"

    def test_lwa_header_takes_priority_over_proxy(self) -> None:
        ctx = {
            "identity": {
                "clientCert": {
                    "subjectDN": "CN=from-lwa,O=Supervictor",
                }
            }
        }
        response = client.get(
            "/",
            headers={
                "x-amzn-request-context": json.dumps(ctx),
                "x-ssl-client-subject-dn": "CN=from-proxy,O=Supervictor",
            },
        )
        assert response.status_code == 200
        assert response.json()["client_subject"] == "CN=from-lwa,O=Supervictor"


class TestExtractClientSubjectNone:
    """No cert headers — local dev behavior."""

    def test_returns_none_without_headers(self) -> None:
        response = client.get("/")
        assert response.status_code == 200
        assert "client_subject" not in response.json()

File: endpoint/tests/test_middleware.rs
use axum::http::HeaderMap;
use supervictor_endpoint::middleware::extract_client_subject;

#[test]
fn extract_from_lwa_header() {
    let mut headers = HeaderMap::new();
    headers.insert(
        "x-amzn-request-context",
        r#"{"identity":{"clientCert":{"subjectDN":"CN=device1,O=supervictor"}}}"#
            .parse()
            .unwrap(),
    );
    let subject = extract_client_subject(&headers);
    assert_eq!(subject.as_deref(), Some("CN=device1,O=supervictor"));
}

#[test]
fn extract_from_proxy_header() {
    let mut headers = HeaderMap::new();
    headers.insert(
        "x-ssl-client-subject-dn",
        "CN=device2,O=supervictor".parse().unwrap(),
    );
    let subject = extract_client_subject(&headers);
    assert_eq!(subject.as_deref(), Some("CN=device2,O=supervictor"));
}

#[test]
fn lwa_takes_priority_over_proxy() {
    let mut headers = HeaderMap::new();
    headers.insert(
        "x-amzn-request-context",
        r#"{"identity":{"clientCert":{"subjectDN":"CN=from-lwa"}}}"#
            .parse()
            .unwrap(),
    );
    headers.insert("x-ssl-client-subject-dn", "CN=from-proxy".parse().unwrap());
    let subject = extract_client_subject(&headers);
    assert_eq!(subject.as_deref(), Some("CN=from-lwa"));
}

#[test]
fn no_headers_returns_none() {
    let headers = HeaderMap::new();
    assert!(extract_client_subject(&headers).is_none());
}

#[test]
fn malformed_lwa_json_returns_none() {
    let mut headers = HeaderMap::new();
    headers.insert("x-amzn-request-context", "{bad json}".parse().unwrap());
    assert!(extract_client_subject(&headers).is_none());
}

#[test]
fn lwa_missing_cert_falls_through() {
    let mut headers = HeaderMap::new();
    headers.insert(
        "x-amzn-request-context",
        r#"{"identity":{}}"#.parse().unwrap(),
    );
    assert!(extract_client_subject(&headers).is_none());
}

#[test]
fn lwa_missing_cert_falls_through_to_proxy() {
    let mut headers = HeaderMap::new();
    headers.insert(
        "x-amzn-request-context",
        r#"{"identity":{}}"#.parse().unwrap(),
    );
    headers.insert("x-ssl-client-subject-dn", "CN=fallback".parse().unwrap());
    let subject = extract_client_subject(&headers);
    assert_eq!(subject.as_deref(), Some("CN=fallback"));
}
